buildSrc/src/main/groovy/net/minecraftforge/forge/tasks/checks/CheckATs.groovy

```diff
@@ -0,0 +1,304 @@
+package net.minecraftforge.forge.tasks.checks
+
+import groovy.transform.CompileStatic
+import groovy.transform.TupleConstructor
+import net.minecraftforge.forge.tasks.InheritanceData
+import net.minecraftforge.srgutils.IMappingFile
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.RegularFile
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.tasks.InputFile
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Optional
+import org.objectweb.asm.Opcodes
+
+@CompileStatic
+abstract class CheckATs extends CheckTask {
+
+    @InputFile abstract RegularFileProperty getInheritance()
+    @InputFiles abstract ConfigurableFileCollection getAts()
+    @InputFile @Optional abstract RegularFileProperty getMappings()
+
+    @Override
+    void check(Reporter reporter, boolean fix) {
+        final IMappingFile mappings = mappings.map { RegularFile it -> IMappingFile.load(it.asFile) }.getOrNull()
+        final inheritance = InheritanceData.parse(this.inheritance.get().asFile)
+
+        ats.each {
+            final lines = process(it, reporter, inheritance)
+            if (fix) {
+                it.text = joinBack(lines, inheritance, mappings).join('\n')
+            }
+        }
+    }
+
+    private static TreeMap<String, ATParser.Entry> process(File file, Reporter reporter, Map<String, InheritanceData> inheritance) {
+        final TreeMap<String, ATParser.Entry> lines = ATParser.parse(file.readLines(), reporter)
+
+        final itr = lines.entrySet().iterator()
+        final toRemove = []
+        while (itr.hasNext()) {
+            final next = itr.next()
+            String key = next.key
+            final entry = next.value
+            if (entry === null) continue
+
+            // Process Groups, this will remove any entries outside the group that is covered by the group
+            if (entry.group) {
+                final jcls = inheritance[entry.cls.replaceAll('\\.', '/')]
+                if (jcls === null) {
+                    itr.remove()
+                    reporter.report("Invalid group: $key")
+                } else if ('*' == entry['desc']) {
+                    if (!jcls.fields) {
+                        itr.remove()
+                        reporter.report("Invalid group, class has no fields: $key")
+                    } else {
+                        jcls.fields.each { field, value ->
+                            final fkey = entry.cls + ' ' + field
+                            if (accessLevel(value.access) < accessStr(entry.modifier)) {
+                                if (lines.containsKey(fkey)) {
+                                    toRemove.add(fkey)
+                                } else if (!entry.existing.contains(fkey)) {
+                                    println('Added: ' + fkey)
```

Comment:
Shouldn't this be reported, given that this means that the check task found a line which _should_ exist in the AT but actually doesn't? (Same with the other nearby `println('Added: ' + fKey)` and `println('Removed: ' + fKey)`.)